File: app/src/main/java/com/nwe/spadesscore/SpadesAppCompatActivity.java
package com.nwe.spadesscore;

import android.os.Bundle;

import androidx.activity.OnBackPressedCallback;
import androidx.appcompat.app.AppCompatActivity;

public abstract class SpadesAppCompatActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initContentView();
        initializeUIComponents();
        setupUI();
        setupBackPressHandler();
    }

    private void setupBackPressHandler() {
        getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() { /* disable back */ }
        });
    }

    protected abstract void initializeUIComponents();

    protected abstract void setupUI();

    protected abstract void initContentView();

}

File: app/src/main/java/com/nwe/spadesscore/SpadesAppCompatActivity.kt
package com.nwe.spadesscore

import android.os.Bundle
import android.view.View
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import com.nwe.spadesscore.ui.applyPersistedLocale
import com.nwe.spadesscore.ui.applySystemBarInsetsAsPadding

abstract class SpadesAppCompatActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        applyPersistedLocale()
        initContentView()
        findViewById<View>(android.R.id.content).applySystemBarInsetsAsPadding()
        initializeUIComponents()
        setupUI()
        setupBackPressHandler()
    }

    private fun setupBackPressHandler() {
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() { /* disable back */ }
        })
    }

    protected abstract fun initializeUIComponents()

    protected abstract fun setupUI()

    protected abstract fun initContentView()
}
